File: ios/MindsetForgeWidget/WidgetPayload.swift
import Foundation

/// Mirror of the Dart `WidgetPayload` (lib/models/widget_payload.dart), decoded
/// from the JSON blob the app writes into the shared App Group via `home_widget`.
///
/// The next-action fields (`sessionLabel`, `headline`, `subline`, `accentKind`,
/// ...) are resolved on the Dart side by the shared `resolveHeroAction`, so the
/// widget shows exactly the same "right now" action as the in-app hero.
struct WidgetPayload: Codable {
    var state: String
    var focusText: String
    var focusDate: String
    var focusCompleted: Bool
    var hasFocusToday: Bool
    var sessionPeriod: String

    // Resolved next-action (mirrors the dashboard hero).
    var actionField: String
    var sessionLabel: String
    var headline: String
    var subline: String
    var accentKind: String
    var canCompleteInWidget: Bool
    var deepLink: String

    var streak: Int
    var completedCount: Int
    var totalCount: Int

    /// Last 7 days of streak qualification (oldest → newest, index 6 = today).
    var weekStreak: [Bool]
    /// Per-day 9/9 perfect flag aligned to `weekStreak` (index 6 = today).
    var weekPerfect: [Bool]
    /// Single-char weekday letters aligned to `weekStreak` (M T W T F S S).
    var weekLabels: [String]
    /// Nudge line shown beneath the 7-day chain in the focus-complete state.
    var weekCaption: String

    var displayName: String
    var firstName: String
    var updatedAt: String

    static let appGroupId = "group.com.mindsetforge.mindsetforge"
    static let payloadKey = "widget_payload"

    static let empty = WidgetPayload(
        state: "set_focus",
        focusText: "",
        focusDate: "",
        focusCompleted: false,
        hasFocusToday: false,
        sessionPeriod: "morning",
        actionField: "setFocus",
        sessionLabel: "TODAY'S FOCUS",
        headline: "Set Your Focus",
        subline: "Choose the #1 action that moves you closest to who you're becoming.",
        accentKind: "set_focus",
        canCompleteInWidget: false,
        deepLink: "mindsetforge://focus",
        streak: 0,
        completedCount: 0,
        totalCount: 8,
        weekStreak: [],
        weekPerfect: [],
        weekLabels: [],
        weekCaption: "",
        displayName: "",
        firstName: "there",
        updatedAt: ""
    )

    /// Reads the latest payload the app stored in the App Group. Falls back to
    /// `.empty` if nothing has been written yet or decoding fails.
    static func load() -> WidgetPayload {
        guard
            let defaults = UserDefaults(suiteName: appGroupId),
            let raw = defaults.string(forKey: payloadKey),
            let data = raw.data(using: .utf8),
            let decoded = try? JSONDecoder().decode(WidgetPayload.self, from: data)
        else {
            return .empty
        }
        return decoded
    }

    // MARK: - Presentation helpers

    /// The calm "complete" states (success accent, no CTA).
    var isDone: Bool { accentKind == "done" }

    /// Whether there's a daily-routine progress value worth rendering.
    var hasProgress: Bool { totalCount > 0 }

    /// Whether a secondary line should render under the headline.
    var hasSubline: Bool { !subline.isEmpty }

    /// Whether a full 7-day streak chain is available to render.
    var hasWeekStreak: Bool { weekStreak.count == 7 && weekLabels.count == 7 }
}

// MARK: - Resilient decoding

// Decoding lives in an extension so the synthesized memberwise initializer
// (used by `.empty`) is preserved. `decodeIfPresent` keeps the widget working
// across version transitions where a stale payload may lack the newer keys.
extension WidgetPayload {
    enum CodingKeys: String, CodingKey {
        case state, focusText, focusDate, focusCompleted, hasFocusToday
        case sessionPeriod, actionField, sessionLabel, headline, subline
        case accentKind, canCompleteInWidget, deepLink
        case streak, completedCount, totalCount
        case weekStreak, weekPerfect, weekLabels, weekCaption
        case displayName, firstName, updatedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        state = try c.decodeIfPresent(String.self, forKey: .state) ?? "set_focus"
        focusText = try c.decodeIfPresent(String.self, forKey: .focusText) ?? ""
        focusDate = try c.decodeIfPresent(String.self, forKey: .focusDate) ?? ""
        focusCompleted = try c.decodeIfPresent(Bool.self, forKey: .focusCompleted) ?? false
        hasFocusToday = try c.decodeIfPresent(Bool.self, forKey: .hasFocusToday) ?? false
        sessionPeriod = try c.decodeIfPresent(String.self, forKey: .sessionPeriod) ?? "morning"
        actionField = try c.decodeIfPresent(String.self, forKey: .actionField) ?? "setFocus"
        sessionLabel = try c.decodeIfPresent(String.self, forKey: .sessionLabel) ?? "TODAY'S FOCUS"
        let decodedHeadline = try c.decodeIfPresent(String.self, forKey: .headline)
        headline = decodedHeadline ?? (focusText.isEmpty ? "Set Your Focus" : focusText)
        subline = try c.decodeIfPresent(String.self, forKey: .subline) ?? ""
        accentKind = try c.decodeIfPresent(String.self, forKey: .accentKind)
            ?? (hasFocusToday ? "focus" : "set_focus")
        canCompleteInWidget = try c.decodeIfPresent(Bool.self, forKey: .canCompleteInWidget)
            ?? (hasFocusToday && !focusCompleted)
        deepLink = try c.decodeIfPresent(String.self, forKey: .deepLink) ?? "mindsetforge://focus"
        streak = try c.decodeIfPresent(Int.self, forKey: .streak) ?? 0
        completedCount = try c.decodeIfPresent(Int.self, forKey: .completedCount) ?? 0
        totalCount = try c.decodeIfPresent(Int.self, forKey: .totalCount) ?? 8
        weekStreak = try c.decodeIfPresent([Bool].self, forKey: .weekStreak) ?? []
        weekPerfect = try c.decodeIfPresent([Bool].self, forKey: .weekPerfect) ?? []
        weekLabels = try c.decodeIfPresent([String].self, forKey: .weekLabels) ?? []
        weekCaption = try c.decodeIfPresent(String.self, forKey: .weekCaption) ?? ""
        displayName = try c.decodeIfPresent(String.self, forKey: .displayName) ?? ""
        firstName = try c.decodeIfPresent(String.self, forKey: .firstName) ?? "there"
        updatedAt = try c.decodeIfPresent(String.self, forKey: .updatedAt) ?? ""
    }
}

File: lib/models/widget_payload.dart
import '../core/constants/app_strings.dart';
import 'daily_completion.dart';
import 'hero_action.dart';
import 'user_profile.dart';

/// Slim, denormalized snapshot of the dashboard hero ("right now" action),
/// shared with the iOS Home/Lock Screen widget (via App Group) and the Apple
/// Watch (via WatchConnectivity).
///
/// Pure data + pure derivation so the native side stays dumb and so the
/// state-resolution logic can be unit-tested without Flutter bindings. The
/// next-action fields come from the shared [resolveHeroAction] so the widget
/// and the in-app `TodayHeroCard` can never drift apart.
class WidgetPayload {
  /// Back-compatible status: `set_focus`, `focus_open`, `on_track`, or the new
  /// `morning` / `evening` routine states.
  final String state;

  /// The user's #1 focus text for today (empty when none set).
  final String focusText;
  final String focusDate; // yyyy-MM-dd
  final bool focusCompleted;
  final bool hasFocusToday;

  /// One of: `morning`, `transition`, `evening`.
  final String sessionPeriod;

  // ── Resolved next-action (mirrors the dashboard hero) ──────────────────────

  /// Win field (`journalCompleted`, `dayPlanned`, ...) or a synthetic id
  /// (`focus` / `setFocus` / `onTrack`). Drives the deep-link target.
  final String actionField;

  /// Uppercase eyebrow label (e.g. `MORNING SESSION`, `TODAY'S FOCUS`).
  final String sessionLabel;

  /// Prominent line shown on the widget (focus text, routine step, or prompt).
  final String headline;

  /// Secondary line under the headline (may be empty).
  final String subline;

  /// Accent bucket the native side maps to a colour:
  /// `morning` | `evening` | `focus` | `set_focus` | `done`.
  final String accentKind;

  /// True only when the action can be completed from the widget (Today's
  /// Focus). Everything else deep-links into the app.
  final bool canCompleteInWidget;

  /// Deep link fired when the widget is tapped.
  final String deepLink;

  final int streak;
  final int completedCount;
  final int totalCount;

  /// Last 7 days of streak qualification (oldest → newest, index 6 = today).
  /// `weekStreak[6]` reflects today's live qualification (5+ of 9 wins).
  final List<bool> weekStreak;

  /// Per-day 9/9 perfect flag aligned to [weekStreak] (index 6 = today).
  /// `perfect` always implies the day also qualifies; the native side renders
  /// these cells with the perfect-day treatment (gradient + star).
  final List<bool> weekPerfect;

  /// Single-char weekday letters aligned to [weekStreak] (e.g. M T W T F S S).
  final List<String> weekLabels;

  /// Nudge line shown beneath the 7-day chain in the focus-complete state.
  final String weekCaption;

  final String displayName;
  final String firstName;
  final String updatedAt; // ISO8601

  const WidgetPayload({
    required this.state,
    required this.focusText,
    required this.focusDate,
    required this.focusCompleted,
    required this.hasFocusToday,
    required this.sessionPeriod,
    required this.actionField,
    required this.sessionLabel,
    required this.headline,
    required this.subline,
    required this.accentKind,
    required this.canCompleteInWidget,
    required this.deepLink,
    required this.streak,
    required this.completedCount,
    required this.totalCount,
    required this.weekStreak,
    required this.weekPerfect,
    required this.weekLabels,
    required this.weekCaption,
    required this.displayName,
    required this.firstName,
    required this.updatedAt,
  });

  /// Empty/sign-in placeholder rendered when no profile is available.
  factory WidgetPayload.empty({DateTime? now}) {
    final at = now ?? DateTime.now();
    return WidgetPayload(
      state: 'set_focus',
      focusText: '',
      focusDate: '',
      focusCompleted: false,
      hasFocusToday: false,
      sessionPeriod: _sessionPeriodFor(at),
      actionField: 'setFocus',
      sessionLabel: AppStrings.heroFocusSessionLabel,
      headline: AppStrings.heroSetFocusLabel,
      subline: AppStrings.heroSetFocusSubtitle,
      accentKind: 'set_focus',
      canCompleteInWidget: false,
      deepLink: 'mindsetforge://focus',
      streak: 0,
      completedCount: 0,
      totalCount: DailyCompletion.totalCount,
      weekStreak: List<bool>.filled(7, false),
      weekPerfect: List<bool>.filled(7, false),
      weekLabels: _weekLabels(at),
      weekCaption: '',
      displayName: '',
      firstName: 'there',
      updatedAt: at.toIso8601String(),
    );
  }

  /// Derives the payload from the live profile and today's completion record.
  /// [now] is injectable for deterministic tests.
  factory WidgetPayload.fromProfile(
    UserProfile profile, {
    DailyCompletion? completion,
    DateTime? now,
  }) {
    final at = now ?? DateTime.now();
    final todayStr = _activeDayString(at);
    final dc = completion ?? _todayCompletion(profile, todayStr);

    final hasFocusToday = profile.dailyFocusAction.isNotEmpty &&
        profile.dailyFocusActionDate == todayStr;
    final focusComplete = profile.isDailyFocusComplete;

    final action = resolveHeroAction(profile, dc, now: at);

    // The widget leads with the actual focus text for the focus state; every
    // other state shows the resolver's title as the headline.
    final isFocusKind = action.kind == HeroActionKind.focus;
    final headline = isFocusKind ? action.subtitle : action.title;
    final subline = isFocusKind ? '' : action.subtitle;

    final weekStreak = _weekStreak(profile, dc, at);
    final weekPerfect = _weekPerfect(profile, dc, at);
    final weekCaption = dc.countsForStreak
        ? AppStrings.widgetStreakSafe
        : '${dc.completedCount}/${DailyCompletion.totalCount} today — '
            '${AppStrings.widgetStreakFinish}';

    return WidgetPayload(
      state: action.legacyState,
      focusText: hasFocusToday ? profile.dailyFocusAction : '',
      focusDate: profile.dailyFocusActionDate,
      focusCompleted: focusComplete,
      hasFocusToday: hasFocusToday,
      sessionPeriod: _sessionPeriodFor(at),
      actionField: action.field,
      sessionLabel: action.sessionLabel,
      headline: headline,
      subline: subline,
      accentKind: action.accentKind,
      canCompleteInWidget: action.actionable,
      deepLink: action.deepLink,
      streak: profile.currentStreak,
      completedCount: dc.completedCount,
      totalCount: DailyCompletion.totalCount,
      weekStreak: weekStreak,
      weekPerfect: weekPerfect,
      weekLabels: _weekLabels(at),
      weekCaption: weekCaption,
      displayName: profile.displayName,
      firstName: profile.firstName,
      updatedAt: at.toIso8601String(),
    );
  }

  static DailyCompletion _todayCompletion(UserProfile profile, String today) {
    return profile.dailyCompletions.firstWhere(
      (c) => c.date == today,
      orElse: () => DailyCompletion(date: today),
    );
  }

  static String _dateString(DateTime d) =>
      '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

  /// The grace-aware active day as a date (midnight–4 AM counts as prior day).
  static DateTime _activeDay(DateTime d) {
    final base = DateTime(d.year, d.month, d.day);
    return d.hour < 4 ? base.subtract(const Duration(days: 1)) : base;
  }

  /// The last 7 active days, oldest → newest (index 6 = today).
  static List<DateTime> _last7Days(DateTime at) {
    final today = _activeDay(at);
    return List.generate(7, (i) => today.subtract(Duration(days: 6 - i)));
  }

  /// Per-day streak qualification for the last 7 days. Today (index 6) uses the
  /// live completion record; prior days look up history. Mirrors the dashboard
  /// `_StreakStrip` so the widget and app never disagree.
  static List<bool> _weekStreak(
    UserProfile profile,
    DailyCompletion today,
    DateTime at,
  ) {
    final days = _last7Days(at);
    final todayKey = _dateString(days.last);
    return days.map((d) {
      final key = _dateString(d);
      if (key == todayKey) return today.countsForStreak;
      final c = profile.dailyCompletions.firstWhere(
        (x) => x.date == key,
        orElse: () => DailyCompletion(date: key),
      );
      return c.countsForStreak;
    }).toList();
  }

  /// Per-day 9/9 perfect status for the last 7 days. Today (index 6) uses the
  /// live completion record; prior days look up history. Mirrors the dashboard
  /// `_StreakStrip` perfect-day branch so the widget and app never disagree.
  static List<bool> _weekPerfect(
    UserProfile profile,
    DailyCompletion today,
    DateTime at,
  ) {
    final days = _last7Days(at);
    final todayKey = _dateString(days.last);
    return days.map((d) {
      final key = _dateString(d);
      if (key == todayKey) return today.isPerfectDay;
      final c = profile.dailyCompletions.firstWhere(
        (x) => x.date == key,
        orElse: () => DailyCompletion(date: key),
      );
      return c.isPerfectDay;
    }).toList();
  }

  /// Single-char weekday letters aligned to [_last7Days]. Locale-independent
  /// (DateTime.weekday: 1 = Mon … 7 = Sun) so it stays in sync with native.
  static List<String> _weekLabels(DateTime at) {
    const letters = ['M', 'T', 'W', 'T', 'F', 'S', 'S']; // Mon → Sun
    return _last7Days(at).map((d) => letters[d.weekday - 1]).toList();
  }

  /// The "active day" key (4 AM–4 AM): midnight–4 AM counts as the prior day so
  /// the widget's completion count and focus check don't reset at midnight
  /// while the session period is still evening. Mirrors
  /// `AppDateUtils.todayStringWithGracePeriod`.
  static String _activeDayString(DateTime d) =>
      _dateString(d.hour < 4 ? d.subtract(const Duration(days: 1)) : d);

  /// Matches `AppDateUtils.sessionPeriod`: morning 4–11, transition 12–16,
  /// evening otherwise (17–23 and 0–3).
  static String _sessionPeriodFor(DateTime d) {
    final hour = d.hour;
    if (hour >= 4 && hour < 12) return 'morning';
    if (hour >= 12 && hour < 17) return 'transition';
    return 'evening';
  }

  /// Convenience for the "complete" states (no CTA, success accent).
  bool get isDone => accentKind == 'done';

  Map<String, dynamic> toJson() => {
        'state': state,
        'focusText': focusText,
        'focusDate': focusDate,
        'focusCompleted': focusCompleted,
        'hasFocusToday': hasFocusToday,
        'sessionPeriod': sessionPeriod,
        'actionField': actionField,
        'sessionLabel': sessionLabel,
        'headline': headline,
        'subline': subline,
        'accentKind': accentKind,
        'canCompleteInWidget': canCompleteInWidget,
        'deepLink': deepLink,
        'streak': streak,
        'completedCount': completedCount,
        'totalCount': totalCount,
        'weekStreak': weekStreak,
        'weekPerfect': weekPerfect,
        'weekLabels': weekLabels,
        'weekCaption': weekCaption,
        'displayName': displayName,
        'firstName': firstName,
        'updatedAt': updatedAt,
      };

  factory WidgetPayload.fromJson(Map<String, dynamic> json) => WidgetPayload(
        state: json['state'] as String? ?? 'set_focus',
        focusText: json['focusText'] as String? ?? '',
        focusDate: json['focusDate'] as String? ?? '',
        focusCompleted: json['focusCompleted'] as bool? ?? false,
        hasFocusToday: json['hasFocusToday'] as bool? ?? false,
        sessionPeriod: json['sessionPeriod'] as String? ?? 'morning',
        actionField: json['actionField'] as String? ?? 'setFocus',
        sessionLabel: json['sessionLabel'] as String? ?? '',
        headline: json['headline'] as String? ?? '',
        subline: json['subline'] as String? ?? '',
        accentKind: json['accentKind'] as String? ?? 'set_focus',
        canCompleteInWidget: json['canCompleteInWidget'] as bool? ?? false,
        deepLink: json['deepLink'] as String? ?? 'mindsetforge://focus',
        streak: (json['streak'] as num?)?.toInt() ?? 0,
        completedCount: (json['completedCount'] as num?)?.toInt() ?? 0,
        totalCount:
            (json['totalCount'] as num?)?.toInt() ?? DailyCompletion.totalCount,
        weekStreak: (json['weekStreak'] as List?)
                ?.map((e) => e as bool? ?? false)
                .toList() ??
            const [],
        weekPerfect: (json['weekPerfect'] as List?)
                ?.map((e) => e as bool? ?? false)
                .toList() ??
            const [],
        weekLabels: (json['weekLabels'] as List?)
                ?.map((e) => e as String? ?? '')
                .toList() ??
            const [],
        weekCaption: json['weekCaption'] as String? ?? '',
        displayName: json['displayName'] as String? ?? '',
        firstName: json['firstName'] as String? ?? 'there',
        updatedAt: json['updatedAt'] as String? ?? '',
      );
}
